plonky3/src/circuit_builder.rs

```diff
@@ -232,12 +240,34 @@ impl<'a, T: FieldElement, AB: AirBuilderWithPublicValues<F = Val>> Air<AB> for P
 
         // public constraints
         let pi_moved = pi.to_vec();
-        let local = matrix.row_slice(0);
+        let (local, next) = (matrix.row_slice(0), matrix.row_slice(1));
 
-        // constraining Pi * (Ci - pub[i]) = 0
         publics.iter().zip(pi_moved).enumerate().for_each(
-            |(index, ((_, col_id, _), public_value))| {
-                let selector = local[self.analyzed.commitment_count() + index];
+            |(index, ((_, col_id, row_id), public_value))| {
+                let (incrementor, inverse, selector) = (
+                    local[self.analyzed.commitment_count() + 3 * index],
+                    local[self.analyzed.commitment_count() + 3 * index + 1],
+                    local[self.analyzed.commitment_count() + 3 * index + 2],
+                );
+
+                //set incrementors for each column to be row_id and decrement each row
+                let incrementor_next = next[self.analyzed.commitment_count() + 3 * index];
+
+                let mut when_first_row = builder.when_first_row();
+                when_first_row.assert_eq(
+                    incrementor,
+                    AB::Expr::from(cast_to_goldilocks(GoldilocksField::from(*row_id as u32))),
```

Comment:
```suggestion
                    cast_to_goldilocks(GoldilocksField::from(*row_id as u32)).into(),
```

Maybe the `into` is not needed even?